controllers/admin_domains.php

```diff
@@ -630,140 +638,185 @@ public function importPackages()
+                                    $this->DomainsTlds->edit($tld, ['package_id' => $package_id]);
+                                    $set_primary_package = false;
+                                } else {
+                                    $this->Packages->edit(
+                                        $package_id,
+                                        ['status' => 'inactive', 'meta' => (array)$package_info['meta']]
+                                    );
+                                }
 
-                        // Report errors
-                        if (($errors = $this->DomainsTlds->errors())) {
-                            break 2;
+                                // Report errors
+                                if (($errors = $this->Packages->errors()) || ($errors = $this->DomainsTlds->errors())) {
+                                    break 2;
+                                }
+                            }
                         }
 
-                        // Mark this package as the primary for this TLD or mark it as inactive
-                        if ($set_primary_package) {
-                            $this->DomainsTlds->edit($tld, ['package_id' => $package_id]);
-                            $set_primary_package = false;
+                        if ($errors) {
+                            // Set error message
+                            $this->Packages->rollback();
+                            $this->outputAsJson([
+                                'error' => $this->setMessage(
+                                    'error',
+                                    $errors,
+                                    true,
+                                    null,
+                                    false
+                                )
+                            ]);
+
+                            return false;
                         } else {
-                            $this->Packages->edit(
-                                $package_id,
-                                ['status' => 'inactive', 'meta' => (array)$package_info['meta']]
-                            );
+                            // Set success message
+                            $this->outputAsJson([
+                                'message' => $this->setMessage(
+                                    'message',
+                                    Language::_('AdminDomains.!success.packages_imported', true),
+                                    true,
+                                    null,
+                                    false
+                                )
+                            ]);
+                            $this->Packages->commit();
+
+                            return false;
                         }
+                    }
+                case 'check':
+                    // Get the TLDs that are already installed
+                    $installed_tlds = array_values(
```

Comment:
This method won't work unfortunately since TLDs can be imported for multiple modules (meaning the tld may exist but not be finished importing) and TLDs can be overwritten (meaning the TLD may exist but has not been overwritten yet in this import).  You'll need to track the list and progress via the session or company setting.

I made a few edits as well that you'll need to pull in.